### PR TITLE
Fix bazel 0.24 incompatibilities

### LINKED ISSUE #108: numerous incompatibilities with bazel 0.24.0+
It looks like bazel 0.24.0 flipped a number of flags that broke things in this repo:
* `--incompatible_disallow_dict_plus` was enabled by default, resulting in `go_genrule` breaking:
```
ERROR: ... in _go_genrule rule //staging/src/k8s.io/apimachinery/pkg/util/sets:set-gen:
Traceback (most recent call last):
        File ".../k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/util/sets/BUILD", line 25
                _go_genrule(name = 'set-gen')
        File ".../external/io_k8s_repo_infra/defs/go.bzl", line 82, in _go_genrule_impl
                ctx.action(inputs = (list(all_srcs) + resol...), <5 more arguments>)
        File ".../external/io_k8s_repo_infra/defs/go.bzl", line 85, in ctx.action
                ctx.configuration.default_shell_env + go.env
The `+` operator for dicts is deprecated and no longer supported. Please use the `update` method instead. You can temporarily enable the `+` operator by passing the flag --incompatible_disallow_dict_plus=fals
```
* The `pkg_tar` Starlark implementation seems to be [including `-mtime` now](https://github.com/bazelbuild/bazel/commit/25d202fc1545f4aaba821df75a8bf3fb57284da1):
```
flag provided but not defined: -mtime
Usage of bazel-out/host/bin/external/io_k8s_repo_infra/tools/build_tar/linux_amd64_stripped/build_tar:
  -compression gz
        Compression (gz or `bz2`), default is none.
  -deb value
        A debian package to add to the layer
  -directory string
        Directory in which to store the file inside the layer
  -file value
        A file to add to the layer
  -flagfile string
        Path to flagfile
  -link value
        Add a symlink a inside the layer ponting to b if a:b is specified
  -mode string
        Force the mode on the added files (in octal).
  -modes value
        Specific mode to apply to specific file (from the file argument), e.g., path/to/file=0455.
  -output string
        The output file, mandatory
  -owner string
        Specify the numeric default owner of all files, e.g., 0.0 (default "0.0")
  -owner_name string
        Specify the owner name of all files, e.g. root.root.
  -owner_names value
        Specify the owner names of individual files, e.g. path/to/file=root.root.
  -owners value
        Specify the numeric owners of individual files, e.g. path/to/file=0.0.
  -tar value
        A tar file to add to the layer
Target //build/release-tars:release-tars failed to build
```
  (See also https://github.com/kubernetes/repo-infra/issues/100)
* The `cc_toolchain` rule seems to have changed its attributes; trying to cross-compile gives lots of errors about missing attributes. (Might be related to https://github.com/bazelbuild/bazel/issues/6861?)

There might be other issues, these were just the first ones I noticed.

### PR DESCRIPTION
**Pull request by ixdy**
Several `--incompatible` flags were flipped in bazel 0.24, causing a number of things to break as noted in #108.

I made some additional changes to `go_genrule` and the `cc_toolchain` target anticipating breaking changes coming in bazel 0.25+. At least a few of these preemptive changes require bazel 0.23+. (Thankfully, I think nothing was broken before bazel 0.24, so we can safely update CI to bazel 0.23 first.)

Fixes #108.
x-ref #100.

/assign @fejta @mikedanese

### CODE REVIEW

File: defs/go.bzl [hunk 1]
```diff
@@ -16,34 +16,34 @@ load("@io_bazel_rules_go//go:def.bzl", "GoLibrary", "GoPath", "go_context", "go_
 
 def _compute_genrule_variables(resolved_srcs, resolved_outs, dep_import_paths):
     variables = {
-        "SRCS": cmd_helper.join_paths(" ", resolved_srcs),
```
Comment (ixdy) [inline defs/go.bzl]: `cmd_helper` is deprecated

File: defs/go.bzl [hunk 2]
```diff
@@ -70,21 +75,23 @@ def _go_genrule_impl(ctx):
...
+            all_srcs.to_list(),
+            ctx.outputs.outs,
             dep_import_paths,
         ),
         tools = ctx.attr.tools,
         label_dict = label_dict,
     )
 
-    paths = ["/bin", "/usr/bin"]
-    ctx.action(
```
Comment (ixdy) [inline defs/go.bzl]: `ctx.action` is deprecated

File: defs/go.bzl [hunk 3]
```diff
@@ -16,34 +16,34 @@ load("@io_bazel_rules_go//go:def.bzl", "GoLibrary", "GoPath", "go_context", "go_
...
     }
     if len(resolved_srcs) == 1:
-        variables["<"] = list(resolved_srcs)[0].path
+        variables["<"] = resolved_srcs[0].path
     if len(resolved_outs) == 1:
-        variables["@"] = list(resolved_outs)[0].path
+        variables["@"] = resolved_outs[0].path
     return variables
 
 def _go_genrule_impl(ctx):
```
Comment (ixdy) [inline defs/go.bzl]: I've tested this on k/k; most notably, the `zz_generated.openapi.go` file remains unchanged.

File: tools/BUILD.bazel [hunk 4]
```diff
@@ -22,14 +22,10 @@ exports_files(["CROSSTOOL"])
...
     all_files = ":empty",
     compiler_files = ":empty",
-    cpu = cpu,
     dwp_files = ":empty",
-    dynamic_runtime_libs = [":empty"],
     linker_files = ":empty",
     objcopy_files = ":empty",
-    static_runtime_libs = [":empty"],
     strip_files = ":empty",
-    supports_param_files = 1,
```
Comment (ixdy) [inline tools/BUILD.bazel]: this is the default

File: tools/BUILD.bazel [hunk 5]
```diff
@@ -22,14 +22,10 @@ exports_files(["CROSSTOOL"])
     name = "cc-gcc-" + cpu,
     all_files = ":empty",
     compiler_files = ":empty",
-    cpu = cpu,
```
Comment (ixdy) [inline tools/BUILD.bazel]: deprecated (I think it went optional in bazel 0.23)

**Comment by ixdy** [comment]
/hold

I like @fejta's fixes to kazel better than mine.

File: tools/BUILD.bazel [hunk 6]
```diff
@@ -22,14 +22,10 @@ exports_files(["CROSSTOOL"])
...
     dwp_files = ":empty",
-    dynamic_runtime_libs = [":empty"],
```
Comment (ixdy) [inline tools/BUILD.bazel]: the `_runtime_libs` attributes changed names, but they're also optional now, so I just dropped them.

File: defs/go.bzl [hunk 7]
```diff
@@ -16,34 +16,34 @@ load("@io_bazel_rules_go//go:def.bzl", "GoLibrary", "GoPath", "go_context", "go_
...
     go = go_context(ctx)
 
-    all_srcs = depset(go.sdk.libs + go.sdk.srcs + go.sdk.tools + [go.sdk.go])
+    transitive_depsets = []
     label_dict = {}
     go_paths = []
 
     for dep in ctx.attr.srcs:
-        all_srcs += dep.files
```
Comment (ixdy) [inline defs/go.bzl]: this one is due to `--incompatible_depset_union`

File: defs/go.bzl [hunk 8]
```diff
@@ -70,21 +75,23 @@ def _go_genrule_impl(ctx):
         attribute = "cmd",
         expand_locations = True,
         make_variables = _compute_genrule_variables(
-            all_srcs,
```
Comment (ixdy) [inline defs/go.bzl]: `--incompatible_depset_is_not_iterable`

File: defs/go.bzl [hunk 9]
```diff
@@ -70,21 +75,23 @@ def _go_genrule_impl(ctx):
...
-        inputs = list(all_srcs) + resolved_inputs,
+    env = {}
+    env.update(ctx.configuration.default_shell_env)
+    env.update(go.env)
+    env["PATH"] = ctx.configuration.host_path_separator.join(["/bin", "/usr/bin"])
+
+    ctx.actions.run_shell(
+        inputs = depset(resolved_inputs, transitive = [all_srcs]),
         outputs = ctx.outputs.outs,
-        env = ctx.configuration.default_shell_env + go.env + {
```
Comment (ixdy) [inline defs/go.bzl]: `--incompatible_disallow_dict_plus`

**Comment by fejta** [comment]
LGTM otherwise

**Comment by ixdy** [comment]
I put these all into one PR because otherwise CI wouldn't pass.

**Comment by fejta** [comment]
> I put these all into one PR because otherwise CI wouldn't pass.

Not technically true but close enough

**Comment by ixdy** [comment]
why did you `/hold cancel` this?

**Comment by fejta** [comment]
Because things don't build, and we need to make forward progress

**Comment by fejta** [comment]
This unblocks merging the other PRs

**Comment by ixdy** [comment]
I was going to rebase to pull in your #113, but whatever.

**Comment by fejta** [comment]
I'm way more interested in finally getting off travis and migrating to go modules than merging my PR, when we already have one that works and has a chance of merging.

**Pull request merged**